Block Bindings: Docs API reference.

cbravobernal opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Initial draft for Block Bindings
Fixes #63026

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Documentation is necessary.

cbravobernal committed:
Initial commit

cbravobernal committed:
Second commit, still not complete

cbravobernal committed:
More docs

cbravobernal committed:
More fixes

cbravobernal committed:
First draft written, still needs grammar check

cbravobernal committed:
Applied AI corrections

justintadlock commented on docs/reference-guides/block-api/block-bindings.md:
````diff
@@ -0,0 +1,296 @@
+```
+
+
+## Compatible blocks and its attributes
````
Should be: Compatible blocks and their attributes

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/b553fcd81feba6985e24825ec252bc0aee2585fc

justintadlock commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+
+## Compatible blocks and its attributes
+
+Right now, not all blocks attributes are compatible with block bindings. There is some effort going on increasing this compatibility, but for now, this is the list:
```
"blocks attributes" should be "block attributes"

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/b553fcd81feba6985e24825ec252bc0aee2585fc

justintadlock commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+
+## Registering a custom source
+
+Registering a source consists of defining a **name** for that source and a callback function specifying how to get a value from that source and pass it to a block attribute.
```
Try:

Registering a source requires defining at least `name` and a callback function that gets a value from the source and passes it back to a block attribute.

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/b553fcd81feba6985e24825ec252bc0aee2585fc

cbravobernal committed:
Changes requested

justintadlock commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+					}
+					return $visualization_date;
+				},
+				'uses_context'       => array( 'postId', 'postType' ),
```
Maybe this is a minor thing, but I don't see the post type used. Do we need `postType` here?

cbravobernal replied:
Fixed in https://github.com/WordPress/gutenberg/pull/66251/commits/f578dbcf95fdcbda2c0cb1f8b2e2238506e1e13e

justintadlock commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+				'show_in_rest'      => true,
+				'single'            => true,
+				'type'              => 'string',
+				'label'				=> __( 'Movie visualization date' ),
```
Spacing doesn't align with other lines. Actually, it looks like all the lines of the array could have their spacing trimmed back.

cbravobernal replied:
Fixed in https://github.com/WordPress/gutenberg/pull/66251/commits/f578dbcf95fdcbda2c0cb1f8b2e2238506e1e13e
Was a tab thing 😓

cbravobernal committed:
Address more reviews

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+    - `getValues` is a `function` that retrieves the values from the source. (optional)
+    - `setValues` is a `function` that allows updating the values connected to the source. (optional)
+    - `canUserEditValue` is a `function` to determine if the user can edit the value. The user won't be able to edit by default. (optional)
+    - `getFieldsList` is a private `function`. It cannot be used yet by third-party developers. It creates a list for the block bindings UI with post meta.
```
Since `getFieldsList` can in fact be used by third-party developers, should we say something like the following instead?

> `getFieldsList` is an experimental `function` that should not be used yet by third-party developers in production, as it is subject to change and may be removed in upcoming releases. It creates a list for the block bindings UI with post meta.

cbravobernal replied:
Fixed in https://github.com/WordPress/gutenberg/pull/66251/commits/2c19ec0a78b180f72d8c55e502cf361095c06354

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+import {
+	registerBlockBindingsSource,
+} from '@wordpress/blocks';
+import { __ } from '@wordpress/i18n';
```
We should add the following so the snippet is functional:

`import { store as coreDataStore } from '@wordpress/core-data';`

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+		);
+	},
+	getValues( { select, context } ) {
+		let wpModiesVisualizationDate;
```
The variable name is misspelled — it should be `wpModiesVisualizationDate`

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+The `getValues` function retrieves the value from the source on block loading. It receives an `object` as an argument with the following properties:
+
+- `bindings` return the bindings object. It must have the attributes as a key, and the value can be a string or an object with arguments.
+- `cliendId` returns a `string` with the current block client ID.
```
`clientId` is misspelled.

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+		let wpModiesVisualizationDate;
+		const { getEditedEntityRecord } = select( coreDataStore );
+		if ( context?.postType && context?.postId ) {
+			return getEditedEntityRecord(
+				'postType',
+				context?.postType,
+				context?.postId
+			).meta?.wp_movies_visualization_date;
```
I think you mean to set `wpMoviesVisualizationDate` rather than return here, otherwise the variable is always undefined.

cbravobernal replied:
Yep. It should be 
```
wpMoviesVisualizationDate = getEditedEntityRecord(
				'postType',
				context?.postType,
				context?.postId
			).meta?.wp_movies_visualization_date;
```

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+registerBlockBindingsSource( {
+	name: 'wpmovies/visualization-date',
+	label: __( 'Visualization Date' ),
+	useContext: [ 'postId', 'postType' ],
```
`usesContext` is misspelled.

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+    - text
+    - url
+    - linkTarget
+    - rel
```
I think this would be better as a table:

| Supported Blocks    | Supported Attributes |
| -------- | ------- |
| Paragraph  | content    |
| Heading | content     |
| Image    | ID, url, title, alt    |
| Button    | text, url, linkTarget, rel    |

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+# Bindings
+
+Block Bindings API lets you “bind” dynamic data to the block’s attributes, which are then reflected in the final HTML markup that is output to the browser on the front end.
```
Missing "the":

> The Block Bindings API...

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+
+## Compatible blocks and their attributes
+
+Right now, not all block attributes are compatible with block bindings. There is some effort going on increasing this compatibility, but for now, this is the list:
```
Suggestion:

> This is some ongoing effort to increase this compatibility...

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+
+Registering a source requires defining at least `name` and a `callback` function that gets a value from the source and passes it back to a block attribute.
+
+Once a source is registered, any block that supports the Block Bindings API can use a value from that source by setting its `metadata.bindings` attribute to a value that refers to the source.
```
Suggestion:

> Once a source is registered, any supporting block's `metadata.bindings` attribute can be configured to read a value from that source.

artemiomorales commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,296 @@
+- `args` is an `array` that contains:
+    - `label` is a `string` with the human-readable name of the custom source.
+    - `uses_context` is an `array` with the block context that is passed to the callback (optional).
+    - `get_value_callback` is the `function` that will run on the bound block render function. It accepts three arguments: `source_args`, `block_instance` and `attribute_name`. This value can be overriden with the filter `block_bindings_source_value`.
```
We could try removing "is" to make this cleaner.

- `name`: `string` that sets the unique ID for the custom source.
- `args`: `array` that contains:
    - `label`: `string` with the human-readable name of the custom source.
    - `uses_context`: `array` with the block context that is passed to the callback (optional).
    - `get_value_callback`: `function` that will run on the bound block's render function. It accepts three arguments: `source_args`, `block_instance` and `attribute_name`. This value can be overridden with the filter `block_bindings_source_value`.

artemiomorales reviewed (commented):
I found a few typos; I'll push a commit to fix those and left a few other suggestions.

One thing: The parameter definitions throughout the document are inconsistent and repeat themselves unnecessarily. I suggest revising the definitions [like in this comment ](https://github.com/WordPress/gutenberg/pull/66251#discussion_r1817385298) using either colons (:) or hyphens (-) between the property name and its description to make these easier to read.

cbravobernal committed:
Update getFieldsList expanation

cbravobernal committed:
Fix wrong function

cbravobernal committed:
Fix misspeel

cbravobernal committed:
Recover Artemio's commits

cbravobernal commented:
@artemiomorales I did a force push with `force-with-lease` that should have not deleted your commits, but did 🤦‍♂️ .
I think I re-addressed your comments, but please if you could do a recheck it would be great.

Sorry for the trouble 😞

gziolo commented:
Excellent work putting this reference document together 👏

The new document still needs to be listed in the manifest file to get exposed in the developers portal:

https://github.com/WordPress/gutenberg/blob/trunk/docs/manifest.json

One important aspect is ensuring that every API is annotated with the major WP version it was introduced. It will save a lot of trouble when someone needs to support older versions than WordPress 6.7, which by the way is still not released so some APIs still don't work on production.

gziolo commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,285 @@
+| -------- | ------- |
+| Paragraph  | content    |
+| Heading | content     |
+| Image    | ID, url, title, alt    |
```
The attribute name is different for id:

https://github.com/WordPress/gutenberg/blob/86a3bc9bb3995928b940f20a934d50a1f6f80ed3/packages/block-library/src/image/block.json#L69

```suggestion
| Image    | id, url, title, alt    |
```

cbravobernal replied:
Fixed in https://github.com/WordPress/gutenberg/pull/66251/commits/a2e5a7ac2d1ff2eafbb1688ebf820bc9b20bc268

gziolo commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,285 @@
+					$post_id = $block_instance->context['postId'];
+					$visualization_date = get_post_meta( $post_id, 'wp_movies_visualization_date', true );
+					if ( ! $visualization_date ) {
+						$visualization_date = date("m/d/Y");
```
I recommend running Prettier on this file to ensure WordPress coding styles are followed.

cbravobernal replied:
Fixed in https://github.com/WordPress/gutenberg/pull/66251/commits/a2e5a7ac2d1ff2eafbb1688ebf820bc9b20bc268

gziolo commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,285 @@
+				'show_in_rest' => true,
+				'single'       => true,
+				'type'         => 'string',
+				'label'	       => __( 'Movie visualization date' ),
```
Translations coming from custom code need to have textdomain provided. Like in line 64 for `Visualization Date`.

cbravobernal replied:
Fixed in https://github.com/WordPress/gutenberg/pull/66251/commits/a2e5a7ac2d1ff2eafbb1688ebf820bc9b20bc268

gziolo commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,285 @@
+- `source_args`: Array containing source arguments.
+- `block_instance`: The block instance object.
+- `attribute_name`: The name of the attribute.
+
```
Actually, the example for the custom source `wpmovies/visualization-date` could be simplified to skip the condition for computing the default date. Instead, it could be presented as an example of using `block_bindings_source_value` filter.

cbravobernal replied:
Fixed in https://github.com/WordPress/gutenberg/pull/66251/commits/a2e5a7ac2d1ff2eafbb1688ebf820bc9b20bc268

gziolo commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,285 @@
+
+There are a few examples in Core that can be used as a reference.
+
+- Post Meta. [Source code](https://github.com/WordPress/wordpress-develop/blob/68d2421d758498095449006a055ad8ce5b7a8a40/src/wp-includes/block-bindings/post-meta.php#L59)
+- Pattern overrides. [Source code](https://github.com/WordPress/wordpress-develop/blob/68d2421d758498095449006a055ad8ce5b7a8a40/src/wp-includes/block-bindings/pattern-overrides.php#L36)
+- Twenty Twentyfive theme. [Source code](https://github.com/WordPress/wordpress-develop/blob/68d2421d758498095449006a055ad8ce5b7a8a40/src/wp-content/themes/twentytwentyfive/functions.php#L130)
```
How much useful these references will be after 1-2 years? How they can be linked different so people don't see the outdated files as they were initially coded?

cbravobernal replied:
Maybe we can just point to the files instead of the lines. Something like
```
- Post Meta. [Source code](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/block-bindings/post-meta.php)
- Pattern overrides. [Source code](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/block-bindings/pattern-overrides.php)
- Twenty Twentyfive theme. [Source code](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-content/themes/twentytwentyfive/functions.php)
```

cbravobernal replied:
Fixed in https://github.com/WordPress/gutenberg/pull/66251/commits/a2e5a7ac2d1ff2eafbb1688ebf820bc9b20bc268

gziolo commented on docs/reference-guides/block-api/block-bindings.md:
````diff
@@ -0,0 +1,285 @@
+	})
+}
+```
+
````
The example for removing a connection is included in JSDoc of the helper and can be copied here:

https://github.com/WordPress/gutenberg/blob/86a3bc9bb3995928b940f20a934d50a1f6f80ed3/packages/block-editor/src/utils/block-bindings.js#L60-L61

cbravobernal replied:
Fixed in https://github.com/WordPress/gutenberg/pull/66251/commits/a2e5a7ac2d1ff2eafbb1688ebf820bc9b20bc268

cbravobernal committed:
Edit references

cbravobernal committed:
Update docs

artemiomorales committed:
Fix typos, formatting, omissions

artemiomorales committed:
Add colons to editor registration definition

artemiomorales committed:
Add missing import

artemiomorales committed:
Add code markdown for type

artemiomorales committed:
Tighten language and format types

artemiomorales commented:
> @artemiomorales I did a push with force-with-lease that should have not deleted your commits, but did 🤦‍♂️ .
I think I re-addressed your comments, but please if you could do a recheck it would be great.
>
> Sorry for the trouble 😞

@cbravobernal No worries, I added back the missing bits and added a few more revisions 👍 Feel free to roll back anything you disagree with.

artemiomorales committed:
Remove extraneous 'a', fix typo

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+
+## Compatible blocks and their attributes
+
+Right now, not all block attributes are compatible with block bindings. This is some ongoing effort to increase this compatibility, but for now, this is the list:
```
I think we want to say "**There** is some ongoing effort", right?

cbravobernal replied:
I followed @artemiomorales suggestion:

![Screenshot 2024-10-29 at 11 45 02](https://github.com/user-attachments/assets/88faa394-c593-48f0-967c-9b312e6102ee)

SantosGuillamot replied:
Oh, okay 🙂 When I read it I wondered what we referred to with "this", but maybe it is because I'm not a native speaker.

cbravobernal replied:
Neither I am 😅

artemiomorales replied:
You're both right, it's a typo 😆 The word used should be "there".

artemiomorales replied:
Fixed 👍

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+
+## Registering a custom source
+
+Registering a source requires defining at least `name` and a `callback` function that gets a value from the source and passes it back to a block attribute.
```
If I remember correctly, `label` is also required.

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/0e951d01e967216e7effdff30215d0439b621f14

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+			'wpmovies/visualization-date',
+			array(
+				'label'              => __( 'Visualization Date', 'custom-bindings' ),
+				'get_value_callback' => function ( array $source_args, $block_instance ) {
+					$post_id = $block_instance->context['postId'];
+					$visualization_date = get_post_meta( $post_id, 'wp_movies_visualization_date', true );
+					return $visualization_date;
+				},
```
I wonder if it'd be valuable to use `$source_args` instead of "hardcoding" the 'wp_movies_visualization_date' key in this and the following examples.

cbravobernal replied:
Oh, that's a good one.

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/0e951d01e967216e7effdff30215d0439b621f14

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+	// Prevent the filter to be applied to other sources.
+	if ( $source_args !== 'wpmovies/visualization-date' ) {
+		return $value;
+	}
```
I think we mean to check against `$name` instead of `$source_args` here. If we decide to use `$source_args` as suggested in the previous comment, it should be something like `$source_args[ 'key' ] `

cbravobernal replied:
Done!

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+
+- `args`: `object` with the following structure:
+    - `name`: `string` with the unique and machine-readable name.
+    - `label`: `string` with the human readable name of the custom source. (optional)
+    - `usesContext`: `array` with the block context that the custom source may need. (optional)
```
Maybe it is worth mentioning that these properties shouldn't be defined if they have already been defined in the server. In most of the cases, they aren't needed.

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/0e951d01e967216e7effdff30215d0439b621f14

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+    - `getValues`: `function` that retrieves the values from the source. (optional)
+    - `setValues`: `function` that allows updating the values connected to the source. (optional)
+    - `canUserEditValue`: `function` to determine if the user can edit the value. The user won't be able to edit by default. (optional)
+    - `getFieldsList`: Experimental `function` that should not be used yet by third-party developers in production, as it is subject to change and may be removed in upcoming releases. It creates a list for the block bindings UI with post meta.
```
Asking because I don't know: Should we document experimental APIs? I expect this one to change, so we shouldn't promote its usage.

gziolo replied:
No, it shouldn't be documented as it isn't possible to use it with WordPress 6.7, and isn't considered stable yet.

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+The `label` argument will override the one defined on the server if they are different.
+
+
+This example will show a custom post meta date in the editor and, if it doesn't exist, it will show today's date. The user can edit the value of the date. (Caution: This example does not format the user input as a date—it's only for educational purposes.)
```
I would polish the example to show an "e2e process" of how a source should be registered, first in the server and then improved in the editor. I'm saying this because I can see we are defining a different `usesContext` when in most cases `label` and `usesContext` shouldn't be defined in the client.

cbravobernal replied:
I added a comment (with the option to skip it). But in my humble opinion, documentation should not follow a tutorial's structure. The developer may access only to this part, skipping previous steps.
Link: https://github.com/WordPress/gutenberg/pull/66251/commits/2a6bf13aa356e4e6263ce4e3d251c587e5b4764e

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+			}
+		);
+	},
+	getValues( { select, context } ) {
```
Nit: I'll probably move this one before `setValues` because it is the order we are using to explain the APIs.

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/0e951d01e967216e7effdff30215d0439b621f14

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+
+The `getValues` function retrieves the value from the source on block loading. It receives an `object` as an argument with the following properties:
+
+- `bindings` returns the bindings object. It must have the attributes as a key, and the value can be a `string` or an `object` with arguments.
```
I would specify that they are only the bindings of that specific source, not all bindings.

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/0e951d01e967216e7effdff30215d0439b621f14

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+
+The `setValues` function updates all the values of the source of the block bound. It receives an `object` as an argument with the following properties:
+
+- `bindings` returns the bindings object. It must have the attributes as a key, and the value can be a `string` or an `object` with arguments. This object contains a `newValue` property with the user's input.
```
I would specify that they are only the bindings of that specific source, not all bindings.

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/0e951d01e967216e7effdff30215d0439b621f14

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,305 @@
+
+## Block Bindings Utils
+
+WordPress 6.7 includes a hook with two helpers that allows developers to edit the `metadata.bindings` attribute easily.
```
Should we document that it can receive a `clientId`?

cbravobernal replied:
Done in https://github.com/WordPress/gutenberg/pull/66251/commits/0e951d01e967216e7effdff30215d0439b621f14

cbravobernal committed:
Address Mario's feedback

cbravobernal committed:
Add comments

cbravobernal committed:
Add manifest

artemiomorales committed:
Fix word usage

cbravobernal committed:
I edited the wrong file

SantosGuillamot commented on docs/reference-guides/block-api/block-bindings.md:
```diff
@@ -0,0 +1,316 @@
+    - `canUserEditValue`: `function` to determine if the user can edit the value. The user won't be able to edit by default. (optional)
+    - `getFieldsList`: Experimental `function` that should not be used yet by third-party developers in production, as it is subject to change and may be removed in upcoming releases. It creates a list for the block bindings UI with post meta.
+
+The `label` argument will override the one defined on the server if they are different.
```
Should we move this sentence to the `label` property explanation?

SantosGuillamot commented:
This is looking great 👏 We can keep iterating on it once it is merged.

> One important aspect is ensuring that every API is annotated with the major WP version it was introduced.

They are already documented in the code, but maybe it makes sense to include something in the docs.

gziolo commented on docs/manifest.json:
```diff
@@ -371,6 +371,12 @@
 		"parent": "block-api"
 	},
+	{
+		"title": "Bindings",
```
So great to see it exposed here ❤️

cbravobernal committed:
Add versions, fix label

cbravobernal commented:
> Excellent work putting this reference document together 👏
> 
> The new document still needs to be listed in the manifest file to get exposed in the developers portal:
> 
> https://github.com/WordPress/gutenberg/blob/trunk/docs/manifest.json
> 
> One important aspect is ensuring that every API is annotated with the major WP version it was introduced. It will save a lot of trouble when someone needs to support older versions than WordPress 6.7, which by the way is still not released so some APIs still don't work on production.

Addressed in https://github.com/WordPress/gutenberg/pull/66251/commits/fa3fbe520cf7c12a4452f261fa46c6842b1d5f9d

SantosGuillamot reviewed (approved):
This is looking good to me ☺️ We can work on improvements in follow-up pull requests if needed.
